File: src/main/java/org/blitzortung/android/app/view/components/StatusComponent.java
package org.blitzortung.android.app.view.components;

import android.app.Activity;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import org.blitzortung.android.AlertResultEvent;
import org.blitzortung.android.alarm.AlarmLabel;
import org.blitzortung.android.alarm.AlarmLabelHandler;
import org.blitzortung.android.alarm.AlertEvent;
import org.blitzortung.android.app.R;
import org.blitzortung.android.protocol.Event;
import org.blitzortung.android.protocol.Listener;

public class StatusComponent implements AlarmLabel, Listener {

    private TextView status;

    private TextView warning;

    private ProgressBar progressBar;

    private ImageView errorIndicator;

    private final AlarmLabelHandler alarmLabelHandler;

    public StatusComponent(Activity activity) {
        status = (TextView) activity.findViewById(R.id.status);

        warning = (TextView) activity.findViewById(R.id.warning);

        progressBar = (ProgressBar) activity.findViewById(R.id.progress);
        progressBar.setVisibility(View.INVISIBLE);

        errorIndicator = (ImageView) activity.findViewById(R.id.error_indicator);
        errorIndicator.setVisibility(View.INVISIBLE);

        alarmLabelHandler = new AlarmLabelHandler(this, activity.getResources());
    }

    public void startProgress() {
        progressBar.setVisibility(View.VISIBLE);
        progressBar.setProgress(0);
    }

    public void stopProgress() {
        progressBar.setVisibility(View.INVISIBLE);
        progressBar.setProgress(progressBar.getMax());
    }

    public void indicateError(boolean indicateError) {
        errorIndicator.setVisibility(indicateError ? View.VISIBLE : View.INVISIBLE);
    }

    public void setText(String statusText) {
        status.setText(statusText);
    }

    @Override
    public void setAlarmTextColor(int color) {
        warning.setTextColor(color);
    }

    @Override
    public void setAlarmText(String alarmText) {
        warning.setText(alarmText);
    }

    @Override
    public void onEvent(Event event) {
        if (event instanceof AlertEvent) {
            alarmLabelHandler.apply(
                    event instanceof AlertResultEvent
                            ? ((AlertResultEvent) event).getAlertResult()
                            : null);
        }
    }
}

File: app/src/main/java/org/blitzortung/android/app/view/components/StatusComponent.kt
/*

   Copyright 2015 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.app.view.components

import android.content.Context
import android.view.View
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import org.blitzortung.android.alert.Alarm
import org.blitzortung.android.alert.AlertLabel
import org.blitzortung.android.alert.AlertLabelHandler
import org.blitzortung.android.alert.NoAlarm
import org.blitzortung.android.alert.event.AlertEvent
import org.blitzortung.android.alert.event.AlertResultEvent

class StatusComponent(
    private val warning: TextView,
    private val status: TextView,
    private val progressBar: ProgressBar,
    private val errorIndicator: ImageView,
    context: Context,
) : AlertLabel {
    private val alertLabelHandler: AlertLabelHandler

    val alertEventConsumer: (AlertEvent) -> Unit

    init {
        progressBar.visibility = View.INVISIBLE
        errorIndicator.visibility = View.INVISIBLE

        alertLabelHandler = AlertLabelHandler(this, context)

        alertEventConsumer = { event ->
            if (event is AlertResultEvent && event.alertResult is Alarm)
            alertLabelHandler.apply( event.alertResult )
        }
    }

    fun startProgress() {
        progressBar.apply {
            visibility = View.VISIBLE
            progress = 0
        }
    }

    fun stopProgress() {
        progressBar.apply {
            visibility = View.INVISIBLE
            progress = progressBar.max
        }
    }

    fun indicateError(indicateError: Boolean) {
        errorIndicator.visibility = if (indicateError) View.VISIBLE else View.INVISIBLE
    }

    fun setText(statusText: String) {
        status.text = statusText
    }

    override fun setAlarmTextColor(color: Int) {
        warning.setTextColor(color)
    }

    override fun setAlarmText(alarmText: String) {
        warning.text = alarmText
    }
}
